vignettes/datatable-programming.Rmd

```diff
@@ -204,6 +204,33 @@ DT[filter_col %in% filter_val,
   )]
 ```
 
+### Substitute functions
+
+A tiny clarification may be useful on how we can substitute a function name in an expression.
+Note that providing `outer="sqrt"` (string) vs. `outer=sqrt` (symbol) is very different:
+```{r substitute_fun1, result='hide'}
+DT[, outer(Sepal.Length), env = list(outer="sqrt"), verbose=TRUE]
+#Argument 'j' after substitute: sqrt(Sepal.Length)
+## DT[, sqrt(Sepal.Length)]
+
+DT[, outer(Sepal.Length), env = list(outer=sqrt), verbose=TRUE]
+#Argument 'j' after substitute: .Primitive("sqrt")(Sepal.Length)
+## DT[, .Primitive("sqrt")(Sepal.Length)]
+```
+And while `.Primitive("sqrt")(Sepal.Length)` still works, it is almost never the desired form.
+
+Even more importantly, if the symbol form is meant to be used, then it can, and should, be used directly in the expression, as there is no need for substitution.
+```{r substitute_fun2, result='hide'}
+DT[, sqrt(Sepal.Length)]
+```
+
+If function name to be substituted needs to be namespace-qualified then namespace and function name can be substituted as any other symbol in the expression:
+```{r substitute_fun3, result='hide'}
+DT[, ns::fun(Sepal.Length), env = list(ns="base", fun="sqrt"), verbose=TRUE]
```

Comment:
this one in particular feels quite unnatural to me -- almost never do we want to separate the namespace from its export. I can't think of any use case for that. and it gets us even further from one of my main points about `do.call()` -- using symbols makes the dependency structure of the code much clearer.

let's drop this examplef. I think the above section is a nice addition -- we can revisit the point about namespace-qualified inputs as we gather more organic use cases in the wild.

Reply:
It can as well be `base::fun(col), env=list(fun="sqrt")`
And also `nsfun(col), env=call("::", as.name("base"), as.name("sqrt"))`

Reply:
Or even parse(text="base::sqrt") which I would prefer not to mention in our docs

Reply:
> using symbols makes the dependency structure of the code much clearer.

The problem is that it changes the code, symbols should be used, and should be preferred but then why to use them in env?

Reply:
> nsfun(col), env=call("::", as.name("base"), as.name("sqrt"))
> Or even parse(text="base::sqrt") which I would prefer not to mention in our docs

these days `str2lang("base::sqrt")` would be better and is sort of "accepted"; anyway, now I think you're hitting on something more interesting: `fun(col); env = list(fun = quote(base::sqrt))` is pretty natural and I think solves all of my concerns.

Reply:
@aitap don't really know now without a nice example showing the issue, env is force evaluated, i/j/by after substation should in theory behave as written by hand.